### PR TITLE
DO NOT MERGE introduce black formatter

### PR DESCRIPTION
**Pull request by joelberkeley**
**DO NOT MERGE** without asking. There are external changes that need to happen at the same time as this being merged.

### CODE REVIEW

File: tests/util/misc.py [hunk 1]
```diff
@@ -96,7 +98,16 @@ def various_shapes(*, excluding_ranks: Container[int] = ()) -> FrozenSet[Tuple[i
...
+        (),
+        (0,),
+        (1,),
+        (0, 0),
+        (1, 0),
+        (0, 1),
+        (3, 4),
+        (1, 0, 3),
+        (1, 2, 3),
+        (1, 2, 3, 4, 5, 6),
```
Comment (joelberkeley) [inline tests/util/misc.py:110]: really?

File: tests/unit/test_observer.py [hunk 2]
```diff
@@ -34,47 +34,56 @@ def _sum_with_nan_at_origin(t: tf.Tensor) -> tf.Tensor:
...
+            tf.constant([[-1.0, 0.0], [1.0, 0.0], [0.0, 2.0], [1.0, 3.0]]),
+            Dataset(
+                tf.constant([[-1.0, 0.0], [1.0, 0.0], [0.0, 2.0], [1.0, 3.0]]),
+                tf.constant([[-1.0], [1.0], [2.0], [4.0]]),
+            ),
+        ),
+        (
+            tf.constant([[0.0, 0.0]]),
+            Dataset(tf.zeros([0, 2]), tf.zeros([0, 1])),
+        ),  # only failure points
```
Comment (joelberkeley) [inline tests/unit/test_observer.py:62]: todo move to above?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -70,6 +70,13 @@ Run tests with
 $ tox -e tests
 ```
 
+#### Code formatting
+
+We format all Python code, other than the notebooks, with [black](https://black.readthedocs.io/en/stable/). Black automates a rough approximation of everyone's favourite code formatting style. You may need to run black before pushing changes, with
```
Comment (sam-willis) [inline CONTRIBUTING.md:75]: controversial ;)

**Review by sam-willis — APPROVED** [review]
LGTM

You not going to use isort as well?

**Comment by joelberkeley** [comment]
> LGTM
> 
> You not going to use isort as well?

Could do. Not got that far

**Pull request merged**